extensions/binding/org.eclipse.smarthome.binding.mqtt.automation/src/main/java/org/eclipse/smarthome/automation/module/mqtt/internal/MQTTModuleTypeProvider.java

```diff
@@ -0,0 +1,215 @@
+import org.eclipse.smarthome.config.core.ParameterOption;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+import org.eclipse.smarthome.io.transport.mqtt.MqttBrokerConnection;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * Lists the MQTT system brokers for the user to choose from.
+ *
+ * @author David Graeff - Initial contribution
+ *
+ */
+@NonNullByDefault()
+@Component(immediate = true)
+public class MQTTModuleTypeProvider implements ModuleTypeProvider, MQTTBrokerConnectionDiscoveryParticipant {
+    protected final Set<ProviderChangeListener<ModuleType>> listeners = new HashSet<>();
+
+    @NonNullByDefault({})
+    private MQTTBrokerConnectionDiscoveryService connectionDiscovery;
+
+    private @Nullable ActionType actionType;
+    private @Nullable Locale cachedLocale;
+
+    @Reference
+    protected void setConnectionDiscoveryService(MQTTBrokerConnectionDiscoveryService mqttService) {
+        this.connectionDiscovery = mqttService;
+        mqttService.addBrokersListener(this);
+    }
+
+    protected void unsetConnectionDiscoveryService(MQTTBrokerConnectionDiscoveryService mqttService) {
+        mqttService.removeBrokersListener(this);
+        this.connectionDiscovery = null;
+    }
+
+    @SuppressWarnings("unchecked")
+    @NonNullByDefault({}) // TODO Until the ModuleTypeProvider interface is fixed
+    @Override
+    public ModuleType getModuleType(String UID, Locale locale) {
+        if (PublishActionHandler.MODULE_TYPE_ID.equals(UID)) {
+            return getActionType(locale);
+        } else if (PublishedMessageTriggerHandler.MODULE_TYPE_ID.equals(UID)) {
+            return getTriggerType(locale);
+        } else {
+            return null;
+        }
+    }
+
+    @NonNullByDefault({}) // TODO Until the ModuleTypeProvider interface is fixed
+    @Override
+    public Collection<ModuleType> getModuleTypes(Locale locale) {
+        cachedLocale = locale;
+        return Stream.of(getActionType(locale), getTriggerType(locale)).collect(Collectors.toList());
+    }
+
+    private ModuleType getActionType(Locale locale) {
+        final ActionType actionType = new ActionType(PublishActionHandler.MODULE_TYPE_ID, getConfigActionDesc(locale), //
+                "Publish MQTT message", "Publishes a message to an MQTT topic", //
+                null, Visibility.VISIBLE, getInputActionDesc(locale), getOutputTriggerDesc(locale));
+        this.actionType = actionType;
+        return actionType;
+    }
+
+    private ModuleType getTriggerType(Locale locale) {
+        return new TriggerType(PublishedMessageTriggerHandler.MODULE_TYPE_ID, getConfigTriggerDesc(locale),
+                "MQTT message published", "Triggers whenever the given MQTT topic has a new value", //
+                null, Visibility.VISIBLE, getOutputTriggerDesc(locale));
+    }
+
+    private @Nullable List<Output> getOutputTriggerDesc(Locale locale) {
```

Comment:
broker & topicName are both static configuration and thus should not be an output. The only "dynamic" information is imho the topicValue, which qualifies as an output.

Reply:
For "broker": That depends on what you think about my suggestion below.
For "topicName": The user can specify a wildcard topic "sensors/+/config" for a trigger. The "topicName" is the actual topic that was received like "sensors/abc/config".